tests/integration/test_engine_integration.py

```diff
@@ -96,84 +91,82 @@ class TestProperExecution:
 
     def test_regref_transform(self, setup_eng):
         """Test circuit with regref transforms doesn't raise any errors"""
-        eng, q = setup_eng(2)
+        eng, prog = setup_eng(2)
 
-        with eng:
+        with prog.context as q:
             ops.MeasureX | q[0]
             ops.Sgate(q[0]) | q[1]
             # symbolic hermitian conjugate together with register reference
             ops.Dgate(q[0]).H | q[1]
             ops.Sgate(ops.RR(q[0], lambda x: x ** 2)) | q[1]
             ops.Dgate(ops.RR(q[0], lambda x: -x)).H | q[1]
 
-        eng.run()
+        eng.run(prog)
 
     def test_homodyne_measurement_vacuum(self, setup_eng, tol):
         """MeasureX and MeasureP leave the mode in the vacuum state"""
-        eng, q = setup_eng(2)
-
-        with eng:
+        eng, prog = setup_eng(2)
+        with prog.context as q:
             ops.Coherent(a, c) | q[0]
             ops.Coherent(b, c) | q[1]
             ops.MeasureX | q[0]
             ops.MeasureP | q[1]
 
-        eng.run()
+        eng.run(prog)
         assert np.all(eng.backend.is_vacuum(tol))
 
     def test_homodyne_measurement_vacuum_phi(self, setup_eng, tol):
         """Homodyne measurements leave the mode in the vacuum state"""
-        eng, q = setup_eng(2)
-
-        with eng:
+        eng, prog = setup_eng(2)
+        with prog.context as q:
             ops.Coherent(a, b) | q[0]
             ops.MeasureHomodyne(c) | q[0]
 
-        eng.run()
+        eng.run(prog)
         assert np.all(eng.backend.is_vacuum(tol))
 
     def test_program_subroutine(self, setup_eng, tol):
         """Simple quantum program with a subroutine and references."""
-        eng, q = setup_eng(2)
+        eng, prog = setup_eng(2)
 
         # define some gates
         D = ops.Dgate(0.5)
         BS = ops.BSgate(0.7 * np.pi, np.pi / 2)
         R = ops.Rgate(np.pi / 3)
-        # get register references
-        alice, bob = q
-
         def subroutine(a, b):
             """Subroutine for the quantum program"""
             R | a
             BS | (a, b)
             R.H | a
 
         # main program
-        with eng:
+        with prog.context as q:
+            # get register references
+            alice, bob = q
             ops.All(ops.Vacuum()) | (alice, bob)
             D | alice
             subroutine(alice, bob)
             BS | (alice, bob)
             subroutine(bob, alice)
 
-        state = eng.run()
+        state = eng.run(prog)
 
         # state norm must be invariant
         if isinstance(eng.backend, BaseFock):
             assert np.allclose(state.trace(), 1, atol=tol, rtol=0)
 
-    def test_checkpoints(self, setup_eng, tol):

```

Comment:
@smite Is this test dropped because it is no longer relevant? i.e., because we no longer have history checkpoints?

Reply:
We do, but they work very differently now. Each Program object stores the state of the RegRefs both at its beginning and at its end. The test no longer tested anything relevant.

Reply:
The Engine only keeps track of the execution history, i.e. the list of Programs executed so far.